### ISSUE TITLE
Double border appears on elements with display: table

### ISSUE THREAD

**Issue by Kelnozz** [servo/servo#11728]
The following code:

```
<!DOCTYPE HTML>
<html>
<head>
<style>
#box {
    display: table;
    border: 1px solid black;
    padding: 10px;
}
</style>
</head>
<div id="box">Hello</div>
</html>
```

Gives that result:

![2016-06-12-112248_661x388_scrot](https://cloud.githubusercontent.com/assets/8824364/15990328/09340f92-3090-11e6-82c1-bb627e2ae260.png)

**Comment by Kelnozz** [comment]
I need to correct myself, It's only a problem when `display: table` is used with `padding` and `border`.

**Comment by mbrubeck** [comment]
Because the table's content is not inside `display: table-row` and `table-cell` elements, Servo creates anonymous rows and cell flows in [`generate_missing_child_flow`][1].  These anonymous flows each get a copy of the table’s CSS styles, which causes the duplicate border and padding in the test case.

I think there are a couple of bugs involved here.

First, according to [CSS2 § 17.6.1][2], borders should not be drawn around table rows:

> Rows, columns, row groups, and column groups cannot have borders (i.e., user agents must ignore the border properties for those elements).

To fix this, [`TableRowFlow::build_display_list`][3] should use `BorderPaintingMode::Hidden` all the time, and not just when `border-collapse` is set to `collapse. Possibly we should also do this in TableRowGroupFlow and TableColGroupFlow.

Second, the anonymous table flows should not "inherit" borders from their parents.  We could fix this by resetting the border to 0 in [`modify_style_for_anonymous_table_object`][4], similar to how we reset padding to 0 in [`modify_style_for_text`][5]. (Or maybe it would be better to have a function that clones only inherited styles, rather than all styles.)

[1]: https://github.com/servo/servo/blob/2086d216ddf2cf87f44dd2da44281daaf810e0a0/components/layout/flow.rs#L1274
[2]: https://drafts.csswg.org/css2/tables.html#separated-borders
[3]: https://github.com/servo/servo/blob/2086d216ddf2cf87f44dd2da44281daaf810e0a0/components/layout/table_row.rs#L424-L435
[4]: https://github.com/servo/servo/blob/2086d216ddf2cf87f44dd2da44281daaf810e0a0/components/style/properties/properties.mako.rs#L2042-L2051
[5]: https://github.com/servo/servo/blob/2086d216ddf2cf87f44dd2da44281daaf810e0a0/components/style/properties/properties.mako.rs#L2078-L2088

**Comment by johannhof** [comment]
I'll take it

**Comment by mbrubeck** [comment]
This is fixed, possibly by #13870.
